### ISSUE TITLE
[feature]: support for trigonometric functions in XLA

### ISSUE THREAD

**Issue by fxsuper** [tensorflow/tensorflow#8315]
I'd like to request support for basic trigonometric functions in XLA, things like cos, sin, etc, as well as basic geometry options like cross products. These are already supported in TF and I imagine there's native support in eigen. Looking through tf2xla I didn't see them anywhere.

**Comment by prb12** [comment]
@learyg @eliben We currently only have Tanh - any reason not to have the rest?

**Comment by fxsuper** [comment]
Any updates on this?

**Comment by eliben** [comment]
I think element-wise trigonometric functions can be added - I don't see a strong reason not to support them, because building them out of other XLA primitives would be challenging

**Comment by fxsuper** [comment]
Does the "contributions welcome" tag mean no one at Google has any plans to add this? Also what about things like `cross`? It's currently a builtin TF function.

**Comment by girving** [comment]
@fxsuper Yep, no one at Google is working on this.

**Comment by learyg** [comment]
Hi! I've just added sin(x) support and it looks like cos(x) support was there. Should be available in the next merge. I'm going to mark this as fixed once that lands, feel free to open new requests for other things, like the cross products you mention, and if possible it'd be helpful to describe why they'd be difficult/impossible to construct with existing XLA operations. (In general it'd also be helpful to have motivating use cases for feature requests, helps us prioritize.) Thanks for filing, sorry for the delay!

**Comment by fxsuper** [comment]
Thanks for the update, and I'm glad that `sin` will be in there soon!

Regarding `cross`--it's certainly possible to build out of existing XLA blocks, but my concern was speed. Is that not a valid concern? TF-level cross can be defined pretty easily indeed:

```
with tf.op_scope([u, v], name, 'cross') as scope:
    u_t = tf.transpose(u)
    v_t = tf.transpose(v)

    u1, u2, u3 = tf.unstack(u_t)
    v1, v2, v3 = tf.unstack(v_t)

    return tf.transpose(tf.stack([(u2 * v3) - (u3 * v2), (u3 * v1) - (u1 * v3), (u1 * v2) - (u2 * v1)]))
```

I'm asking so that I know whether to start a new issue or not. Thanks.

**Comment by learyg** [comment]
Merged as 2661f6841d0ad9ec1381d177a1f9df02e73d001c

Hmm I'm not familiar with how those stack/unstacks will lower from TF to XLA. Could you open a new issue for that and we can check it out, since it's fairly distinct, and address it if there are missing pieces? Thanks much!

**Comment by learyg** [comment]
Also, in reply to #9, looks like it works, but please file the bug to add the op-registered support for tf.cross that does the desugaring you've suggested: https://gist.github.com/learyg/4019bae1cfc5bacbcdb318fd882b4d5d

Looks like it fuses somewhat reasonably, but the CPU backend doesn't seem to currently support fusing the slice operation into its consumers, so it could be better. Thanks for filing!

**Comment by fxsuper** [comment]
Thanks for investigating this! I just opened a new issue.
